### PR TITLE
Feature/typescript

### PR DESCRIPTION
**Pull request by Alexnortung**
### What does it do?

Moved files to typescript and resolved some type issues and changed imports to use `import/export`

### Why is it needed?

I think it is important to have type safety whether it is a plugin or an application. This is especially important in dynamically typed languages such as javascript where runtime errors pop up all the time.

Typescript will also help maintainability of this project by making it easier to implement features and finding bugs if there are inconsistent types.

### How to test it?

run `yarn build` and the plugin should work like before

### Related issue(s)/PR(s)

#31

### CODE REVIEW

File: .eslintrc
Comment (boazpoolman) [inline .eslintrc:77]: These rules should be turned back on.
To prevent issues we should implement eslint overrides config for typescript specific linting rules.
See an example of that here
https://gitlab.uncinc.nl/pub/eslint-config/-/blob/master/index.js#L40

**Review by boazpoolman — CHANGES_REQUESTED** [review]
Other things we should do:

- [x] Update the `CONTRIBUTING.md` with documentation about using the `yarn develop` script
- [ ] We should update the `files` key in the package.json to be able to release the tsc build files to NPM
- [ ] We should update the `eslint` and `eslint:fix` commands in the package.json to be able to lint typescript files as wel

**Review by boazpoolman — APPROVED** [review]

**Review by boazpoolman — CHANGES_REQUESTED** [review]
I would suggest enabling the recommended typescript linting rules.
Just uncomment these in the eslintrc:
```
       // "plugin:@typescript-eslint/eslint-recommended",
        // "plugin:@typescript-eslint/recommended-requiring-type-checking"
```

Splitting up the linting rules for server and admin isn't absolutely required but I can see why you would do that. So you can apply react linting rules only for the admin side. Then again. React rules won't trigger in an nodejs application. I'll leave that up to you to decide.

About the issue with `strapi-admin.js` that cannot find `./admin` is because it's a .js file that tries to import a .ts file. That doesn't work. We could import it from the dist as you did for the server?

File: strapi-server.js
Comment (boazpoolman) [inline strapi-server.js:3]: Github Actions is going on about linting errors for this rule.
It's because the `dist` doesn't exist yet when we run linting in the pipeline.
You could resolve that by adding an extra step in the pipeline that does a build.
Then the linting step should require that to finish before it can run.

**Comment by Alexnortung** [comment]
Before going further, is there a way to easily extend on the rules of the base eslint config? It seems that it is also overriding all the rules since we use

```
"plugin:@typescript-eslint/eslint-recommended",
"plugin:@typescript-eslint/recommended-requiring-type-checking"
```

**Comment by boazpoolman** [comment]
> Before going further, is there a way to easily extend on the rules of the base eslint config? It seems that it is also overriding all the rules since we use

As far as I'm concerned no rules are inherited from the base eslint config.
The overrides basicly start from scratch.

**Comment by Alexnortung** [comment]
I have tested some of the basic functionality on a local project and it seems to work!

Some notable changes:
* warnings from eslint should not be treated as errors and can be fixed along the way
* I have modified some eslint rules so they emit warnings instead of errors
* Eslint has been split into two overrides to accomodate the two tsconfig files 

Would you like to give a final review @boazpoolman ? :)

File: package.json
Comment (boazpoolman) [inline package.json:14]: I'm OK with removing the `--max-warnings=0` for now, but I would suggest turning it back on before the next NPM release.

I know it's a tough measure, but this is an open-source repo and if we allow eslint warnings then people without proper linting config in their IDE could contribute linting warnings without us knowing about it. I really think it's an important one.

**Review by boazpoolman — CHANGES_REQUESTED** [review]
@Alexnortung I still have a couple remarks, but overall this looks very promising!
If you could resolve my last few comments we can merge this.
Good work 💪

**Review by boazpoolman — CHANGES_REQUESTED** [review]
@Alexnortung I still have a couple remarks, but overall this looks very promising!
If you could resolve my last few comments we can merge this.
Good work 💪

File: package.json
Comment (Alexnortung) [reply package.json:14]: I definitely agree that it should be turned back on!
Why should it be before the next npm release?

Comment (boazpoolman) [reply package.json:14]: Just to time-box it, so we won't forget.

Comment (Alexnortung) [reply package.json:14]: We could have a separate workflow that could test only the changed files. I found the following action https://github.com/marketplace/actions/get-all-changed-files that we could use for this purpose.

What do you think of that? :)

Comment (boazpoolman) [reply package.json:14]: I lean towards just running eslint for all the files.
Let's just keep the `--max-warnings=0` off for now and we'll see what happends :)

Comment (Alexnortung) [reply package.json:14]: Awesome! if you are okay with the changes, could you approve them? :)

Comment (boazpoolman) [reply package.json:14]: Yeah, if you fix the build errors I'll approve :)

**Review by boazpoolman — APPROVED** [review]
LGTM! You can merge yourself if you want :)

**Pull request merged**
